File: Templates/Day_3/Vectorisation_SIMD/Fortran/sin_of_vector_serial.f90
Program sin_of_vector

  use trig_taylor

  implicit none

  integer, parameter :: nsize = 8192
  double precision, dimension(nsize) :: x, y
  
  integer :: index


  ! intialise the vector
  Do index = 1, nsize
     x(index) = 0.001D0 * dble(index)
  end do

  ! calculate the sin
  Do index = 1, nsize
     y(index) = sin_taylor(x(index)) 
  end do

  ! write the result
  open(11,file="sin_vector.out")
  do index = 1, nsize
     write(11,*) x(index), y(index)
  end do
  close(11);

  print *, "Calculation finished for", nsize, " elements"

End program sin_of_vector

File: Templates/Archive/Vectorisation_SIMD/C/sin_of_vector_serial.c
#include <stdio.h>
#include <stdlib.h>
#include "trig_taylor.h"

int main()
{

  const int nsize = 8192;

  double* x = aligned_alloc(64, nsize* sizeof *x);
  double* y = aligned_alloc(64, nsize* sizeof *x);

  for (int i=0; i<nsize; i++)
    x[i] = 0.001 * (double)i;

  // calculate the sin
  for (int i=0; i<nsize; i++)
    y[i] = sin_taylor(x[i]);

  // write the result
  FILE * fp;
  fp = fopen("sin_result.out", "w");
  for(int i=0; i<nsize; i++)
    fprintf(fp, "%8.4f %12.9f\n", x[i], y[i]);
  fclose;

  printf("Finished calculation for %d elements\n", nsize);

  free(x);
  free(y);
  return 0;

}
